TS4.0 Variadic Tuple Types (#172)
typescript-kr/typescript-kr.github.io: pages/release-notes/typescript-4.0.md

oneline: TypeScript 4.0 Release Notes
---

> ## 가변 인자 튜플 유형 (Variadic Tuple Types)

Reviewer: ```suggestion
## 가변 인자 튜플 타입 (Variadic Tuple Types)
```

Author: 수정 완료

> 두 개의 배열 혹은 튜플 타입을 연결하여 새로운 배열을 만드는 JavaScript의 `concat` 함수에 대해서 생각해봅시다.

Reviewer: ```suggestion
배열이나 튜플 타입 두 개를 연결하여 새로운 배열을 만드는 JavaScript의 `concat` 함수에 대해서 생각해봅시다.
```

Reviewer: [제안]

Author: 수정 완료

```js
function concat(arr1, arr2) {
  return [...arr1, ...arr2];
}
```

> 또한, 배열이나 튜플을 변수로 입력받아 첫 번째 원소를 제외한 모든 원소를 반환해주는 `tail` 함수에 대해서도 생각해봅시다.

Reviewer: ```suggestion
또한, 배열이나 튜플을 인자로 입력받아 첫 번째 원소를 제외한 모든 원소를 반환해주는 `tail` 함수에 대해서도 생각해봅시다.
```

> 또한, 배열이나 튜플을 인자로 입력받아 첫 번째 원소를 제외한 모든 원소를 반환해주는 `tail` 함수에 대해서도 생각해봅시다.

Reviewer: ```suggestion
그리고 배열이나 튜플을 인자로 입력받아 첫 번째 원소를 제외한 나머지를 반환하는 `tail` 함수에 대해서도 생각해봅시다.
```

Reviewer: [제안]

Author: 수정 완료

function concat<A, B, C, D, E, F>(arr1: [A, B, C, D, E, F], arr2: []): [A, B, C, D, E, F];)
```

> 음... 네, 두 번째의 배열이 항상 비어있는 경우 작성된 7개의 오버로드들입니다.

Reviewer: ```suggestion
음... 네, 이 오버로드들의 두번째 배열은 전부 비어있습니다.
```

Author: 수정 완료

function concat<A1, B1, C1, D1, E1, F1, A2>(arr1: [A1, B1, C1, D1, E1, F1], arr2: [A2]): [A1, B1, C1, D1, E1, F1, A2];
```

> 이것이 지나치다는 것이 분명해지기를 바랍니다.

Reviewer: ```suggestion
이런 오버로딩 함수들은 분명 비합리적입니다.
```
의역을 한번 해봤습니다. [제안]

Author: 이 부분이 되게 애매했었는데 아주 좋은 제안인 것 같습니다. 수정 완료!

> 안타깝게도, `tail`과 같은 함수를 타이핑할 때 이와 같은 비슷한 종류의 문제에 직면하게 될 것입니다.

Reviewer: ```suggestion
안타깝게도, `tail` 함수를 타이핑할 때도 이와 같은 비슷한 문제에 직면하게 됩니다.
```
[제안]

Author: 제안해주신 것 중에 `이와 같은 비슷한 문제` 를 `이와 비슷한 문제`로 수정했습니다. :)

> 이것은 "천 개의 오버로드로 인한 죽음(death by a thousand overloads)"이라고 부르고 싶은 또다른 경우이며, 일반적으로 문제를 해결하지도 않습니다.

Reviewer: ```suggestion
이것은 "천 개의 오버로드로 인한 죽음(death by a thousand overloads)"의 하나의 경우이며, 심지어 대부분 문제를 해결하지도 못합니다.
```
[제안]

Author: 수정 완료

> 우리가 작성하고자하는만큼의 오버로드에 한해서만 올바른 타입을 제공합니다.

Reviewer 2: ```suggestion
우리가 작성하고자 하는 만큼의 오버로드에 한해서만 올바른 타입을 제공합니다.
```

포괄적인 케이스를 만들고 싶다면, 다음과 같은 오버로드가 필요합니다.

```ts
function concat<T, U>(arr1: T[], arr2: U[]): Array<T | U>;
```

> 그러나 해당 시그니처는 튜플을 사용할 때 입력의 길이나 요소의 순서에 대한 어떤 것도 인코딩하지 않습니다.

Reviewer: ```suggestion
그러나 위 시그니처는 튜플을 사용할 때 입력 길이나 요소 순서에 대한 어떤 것도 처리하지 않습니다.
```
[제안]

Author: 수정 완료

> TypeScript 4.0은 추론 개선과 함께 두 가지 핵심적인 변화를 가져와 이러한 타이핑을 가능하게 합니다.

Reviewer: ```suggestion
TypeScript 4.0은 타입 추론 개선을 포함한 두 가지 핵심적인 변화를 도입해 이러한 타이핑을 가능하도록 만들었습니다.
```
[제안]

Author: 수정 완료

> 첫 번째 변화는 튜플 타입 구문의 확장 연산자에서 제너릭 타입을 사용할 수 있다는 점입니다.

Reviewer 2: [제안]
```suggestion
첫 번째 변화는 튜플 타입 구문의 스프레드 연산자에서 제네릭 타입을 사용할 수 있다는 점입니다.
```

> 이 것은 우리가 작동하는 실제 타입을 모르더라도 튜플과 배열에 대해 고차함수를 표현할 수 있다는 뜻입니다.

Reviewer: ```suggestion
우리가 작동하는 실제 타입을 모르더라도 튜플과 배열에 대한 고차함수를 표현할 수 있다는 뜻입니다.
```

Author: 수정 완료

> 이러한 튜플 타입에서 제너릭 확장 연산자가 인스턴스화(혹은, 실제 타입으로 대체)되면 또다른 배열이나 튜플 타입 세트를 생산할 수 있습니다.

Reviewer 3: ```suggestion
이러한 튜플 타입에서 제너릭 확장 연산자가 인스턴스화(혹은, 실제 타입으로 대체)되면 또 다른 배열이나 튜플 타입 세트를 생산할 수 있습니다.
```

> 이러한 튜플 타입에서 제너릭 확장 연산자가 인스턴스화(혹은, 실제 타입으로 대체)되면 또 다른 배열이나 튜플 타입 세트를 생산할 수 있습니다.

Reviewer 2: [제안]
```suggestion
이러한 튜플 타입에서 제너릭 스프레드 연산자가 인스턴스화(혹은, 실제 타입으로 대체)되면 또 다른 배열이나 튜플 타입 세트를 생산할 수 있습니다.
```

const myArray = ["hello", "world"];

const r1 = tail(myTuple);
//    ^ = const r1: [2, 3, 4]

const r2 = tail([...myTuple, ...myArray] as const);
//    ^ = const r2: [2, 3, 4, ...string[]]
```

두 번째 변화는 나머지 요소가 끝뿐만 아니라 튜플의 어느 곳에서도 발생할 수 있다는 것입니다.

```ts
type Strings = [string, string];
type Numbers = [number, number];

type StrStrNumNumBool = [...Strings, ...Numbers, boolean];

> //   ^ = type StrStrNumNumBool = [string, string, number, number, boolean]

Reviewer 2: 👍

}
```

> 하나의 시그니처가 조금 길더라도, 반복할 필요가 없는 하나의 시그니처일뿐이며, 모든 배열과 튜플에서 예측가능한 행동을 제공합니다.

Reviewer 3: ```suggestion
하나의 시그니처가 조금 길더라도, 반복할 필요가 없는 하나의 시그니처일 뿐이며, 모든 배열과 튜플에서 예측 가능한 행동을 제공합니다.
```

이 기능은 그 자체만으로도 훌륭하지만, 조금 더 정교한 시나리오에서도 빛을 발합니다.
예를 들어,[함수의 매개변수를 부분적으로 적용하여 새로운 함수를 반환하는](https://en.wikipedia.org/wiki/Partial_application) `partialCall` 함수가 있다고 생각해봅시다.
`partialCall`은 다음과 같은 함수를 가집니다. - `f`가 예상하는 몇 가지 인수와 함께 `f`라고 지정하겠습니다.

> 그 후, `f`가 여전히 필요로하는 다른 인수를 가지고, 그것을 받을 때 `f`를 호출하는 새로운 함수를 반환합니다.

Reviewer 3: ```suggestion
그 후, `f`가 여전히 필요로 하는 다른 인수를 가지고, 그것을 받을 때 `f`를 호출하는 새로운 함수를 반환합니다.
```

Author: 맞춤법 검사를 다 못했었네요 ㅠㅠ 수정했습니다!

const f2 = partialCall(foo, "hello", 100, true, "oops");

// 작동합니다!
const f3 = partialCall(foo, "hello");

> //    ^ = const f3: (y: number, z: boolean) => void

Reviewer 2: 👍

// f3으로 뭘 할 수 있을까요?

// 작동합니다!
f3(123, true);

f3();

f3(123, "hello");
```

가변 인자 튜플 타입은 특히 기능 구성과 관련하여 많은 새로운 흥미로운 패턴을 가능하게 합니다.
우리는 JavaScript에 내장된 `bind` 메서드의 타입 체킹을 더 잘하기 위해 이를 활용할 수 있을 것이라고 기대합니다.

> 몇 가지 다른 추론 개선 및 패턴들도 여기에 포함되어있으며, 가변 인자 튜플에 대해 더 알아보고 싶다면, [the pull request](https://github.com/microsoft/TypeScript/pull/39094)를 참고해보세요.

Reviewer 2: ```suggestion
몇 가지 다른 추론 개선 및 패턴들도 여기에 포함되어있으며, 가변 인자 튜플에 대해 더 알아보고 싶다면, [the pull request](https://github.com/microsoft/TypeScript/pull/39094)를 참고해보세요.
```
```suggestion
몇 가지 다른 추론 개선 및 패턴들도 여기에 포함되어 있으며, 가변 인자 튜플에 대해 더 알아보고 싶다면, [the pull request](https://github.com/microsoft/TypeScript/pull/39094)를 참고해보세요.
```
